Add Azure Programmable Connectivity dataplane APIs

AndrewNivenMSFT opened the pull request:
# Data Plane API - Pull Request

Introduce new data plane API for Azure Programmable Connectivity.

Azure Programmable Connectivity is a new service going into Public Preview in February. It provides a unified, standard interface across Telecom Operator networks, enabling developers to create cloud and edge-native applications that interact with the intelligence of networks.

## API Info: The Basics

* Link to API Spec engagement record issue: https://github.com/Azure/azure-rest-api-specs-pr/issues/15963

This PR introduces four APIs: three primary APIs, and one auxiliary.

- SimSwap - used by developers to retrieve the time when a SIM card in a device was most recently changed, or verify if the card has changed in a given time frame (useful in fraud detection / prevention)
- Location - used by developers to verify the location of a user
- Number - used by developers to retrieve or verify the phone number of a user
- Network - an auxiliary API used by developers to identify which Network Operator a particular device is connected to. This network is then used in calls to any of the three primary APIs.

For our glossary, core concepts and hero scenarios see [here](https://microsoft.sharepoint.com/:w:/t/CommunicationsSDK/Ea-RP0v4Y1VLjwM8P-jWWdcBh-mepe40JhpZD0xrwlafHg?e=1x3IjB)

Some thoughts and discussions on design can be found [here](https://microsoft.sharepoint.com/:w:/r/teams/CommunicationsSDK/Shared%20Documents/Alpaca%20-%20(Gateway)%20-%20General/API%20and%20SDK%20Review%20Processes/APC%20API%20questions%20.docx?d=we5538ba2edea4887a78ab582719792e9&csf=1&web=1&e=epn5T6) 

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope

n.a. - entirely new API

### Viewing API changes

n.a. - entirely new API

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

<details>
  <summary>Checks stuck in `queued` state?</summary>
If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.  
</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/26925

AndrewNivenMSFT committed:
Move initial work over from private repo

jamie-ford-microsoft committed:
Fix location API - Remove MaxAge parameter

AndrewNivenMSFT committed:
Add some more TODOs

DominikMe commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/number.tsp:
```diff
@@ -0,0 +1,89 @@
+  phoneNumber?: string;
+
+  @doc("Hashed phone number. SHA-256 (in hexadecimal representation) of the mobile phone number in **E.164 format (starting with country code)**. Optionally prefixed with '+'.")
+  hashedPhoneNumber?: string;
```
So you're storing two hashes to compare against - one time with + and one time without?
When would a client send a hashed number versus the plaintext number - is this directly taken from CAMARA?

ellabrmicrosoft replied:
This is taken directly from Camara and its up to the client whether or not they want to use a hashed number or plain text.

ellabrmicrosoft committed:
Resolve TODOs

ellabrmicrosoft committed:
fix pipeline issues

AndrewNivenMSFT commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/network.tsp:
```diff
@@ -0,0 +1,58 @@
+
+interface Networks {
+  @doc("Retrieves the network a given device is on. Returns network in a networkCode format that can be used for other APIs.")
+  retrieve is Operations.ResourceAction<
```
**Question:** One of Dominik's suggestions on your word document was to implement Networks as a resource that we can GET, rather than using a POST retrieve. 

My instinct is that what you've got here is correct, but it would be good to see an explanation (either here or in the word doc) of why we're doing it this way

ellabrmicrosoft replied:
Partly I didn't think about it very much once I decided it was better to stick with what we have than do the headers. I had three reasons for not following this pattern though I think: 
1) This is an API that we might deprecate before GA so I don't know whether its worth putting much effort into it at this point. 
2) Consistency across our APIs feels valuable 
3) This is potentially an error of judgement on my part but I think viewing networks as the resource with IP as the key felt odd to me, I think becuase I think of the network an IP is on as being a feature of the IP address not the network. It feels like an extension of the "retrieving some information about a device" pattern that we have for the other APIs. 

But essentially- I think this way is fine and I think editing this API to be potentially slightly better is very low priority at this point. I'll put it in the excel sheet as something to consider though!

AndrewNivenMSFT replied:
I think that you're right - point 3 is the most important one for me. We don't maintain state about networks, and we don't expect users to "get" them using a known ID, which is what a RESTful design here would imply - as you say, this API is like the others, in that it's a function that accepts a parameter, so it makes sense for it to be a POST as well. It'd be good if you could add this thinking to the document so we record that we've considered that comment.

AndrewNivenMSFT commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/number.tsp:
```diff
@@ -0,0 +1,92 @@
+
+/// Endpoints
+
+@resource("Apis")
```
**Thought:** It seems weird to have four different Endpoint objects with the decorator @resource("Apis"). On the other hand I think this is how we have to do it - if we combine all the Endpoints into one, it's not obvious to me how make it clear which path maps to which API. Do you have any more thoughts about this?

ellabrmicrosoft replied:
You're right, I think its ugly, and its definitely more ugly since setting this up as a proper enum. I've had a little play around with trying to combing the models, but without much success- have got a few things to compile but they don't look like what we want. I'll potentially ask in the typespec channel, or just ask explicitly if we could be doing something better in the API review meeting

ellabrmicrosoft commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/network.tsp:
```diff
@@ -0,0 +1,58 @@
+
+/// Interfaces
+
+interface Networks {
```
No, it means that we are going to allow developers to either make two calls or one call. I put some thoughts about this in the doc here:specification/programmableconnectivity/Azure.ProgrammableConnectivity/examples/LocationInterface_Verify_MaximumSet_Gen.json. If a developer wants to make multiple calls about one device they could get lower latency by using the network API and then using the network code I don't know how significant this will end up being though, so we can always decide to depricate.

AndrewNivenMSFT replied:
Right, yes, sorry, I forgot that the conclusion of that discussion was to keep the Network API as well as supporting passing in IP addresses as a network identifier.

For anyone else reading this, I think this doc is what you meant to link to: https://microsoft.sharepoint.com/:w:/r/teams/CommunicationsSDK/Shared%20Documents/Alpaca%20-%20(Gateway)%20-%20General/API%20and%20SDK%20Review%20Processes/[NetworkAPIProposal.docx](https://microsoft.sharepoint.com/:w:/r/teams/CommunicationsSDK/Shared%20Documents/Alpaca%20-%20(Gateway)%20-%20General/API%20and%20SDK%20Review%20Processes/NetworkAPIProposal.docx?d=wd4bcfc631f054cc1a4676e79795e0599&csf=1&web=1&e=n4kecS)?d=wd4bcfc631f054cc1a4676e79795e0599&csf=1&web=1&e=n4kecS

ellabrmicrosoft committed:
markups

ellabrmicrosoft commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/location.tsp:
```diff
@@ -0,0 +1,87 @@
+  ...PhoneNumberModel;
+
+  @doc("IPv4 address and port of the device")
+  ipv4Address?: Ipv4Address;
```
I don't want to use the spread operator because I want it to be obvious that these are grouped together and keeps the "levels" seperate. If we look at the examples (obvisulsy a user wouldn't use all of these at once.). Without spread:
      "locationDevice": {
        "networkAccessIdentifier": "122345@domain.com",
        "phoneNumber": "+447000000000",
        "ipv4Address": {
          "ipv4": "12.12.12.12",
          "port": 2442
        },
        "ipv6Address": {
          "ipv6": "3001:0da8:75a3:0000:0000:8a2e:0370:7334",
          "port": 1643
        }
      }
and the actual call:

      "locationDevice": {
        "ipv4Address": {
          "ipv4": "12.12.12.12",
          "port": 2442
        },
      }

With spread
      "locationDevice": {
        "networkAccessIdentifier": "122345@domain.com",
        "phoneNumber": "+447000000000",
         "ipv4": "12.12.12.12",
         "port": 2442
         "ipv6": "3001:0da8:75a3:0000:0000:8a2e:0370:7334",
         "port": 1643
      }
and the actual call:

      "locationDevice": {
         "ipv4": "12.12.12.12",
         "port": 2442
      }
      
     It makes it look like "port" is on the same level as "phoneNumber", which it is not. And I think it makes it less obvious what needs to be included when you can't just say you need to include 1 of the top level objects and they need to include whatever they need. Maybe you have a different intuition here?

AndrewNivenMSFT replied:
I agree with your dislike of that second structure. My suggestion included an extra model to provide that extra level of hierarchy - when you spread my suggested `Ipv4AddressModel` into the RequestBody, you should get something that looks like your first structure, rather than your second.

ellabrmicrosoft replied:
I don't know whether I really buy that that is a better structure, seems like an unnecessary layer of abstraction. But I don't think it has real downsides either so will add it

AndrewNivenMSFT replied:
It's just an aesthetic suggestion, it's up to you!

ellabrmicrosoft replied:
Have changed this and the external id for consistency

ellabrmicrosoft committed:
markups

ellabrmicrosoft committed:
markups

ellabrmicrosoft committed:
make extenrla id also consistant with strucutre

ellabrmicrosoft committed:
fix formatting

AndrewNivenMSFT committed:
Add Device to SimSwapVerifyRequest, and make IP addresses optional in LocationVerifyRequest

aronhegedusmicrosoftp committed:
change the scope URL to https://management.azure.com/.default

aronhegedusmicrosoftp committed:
change scopein openapi.json to https://management.azure.com/.default as well

aronhegedusmicrosoftp committed:
Adding python config to tspconfig.yaml

weidongxu-microsoft commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/tspconfig.yaml:
Please add other language config
doc https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/typespec-structure-guidelines.md#structure

weidongxu-microsoft commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/tspconfig.yaml:
```diff
@@ -0,0 +1,21 @@
+    output-file: "{azure-resource-provider-folder}/{service-name}/{version-status}/{version}/openapi.json"
+  "@azure-tools/typespec-python":
+    package-mode: "dataplane"
+    package-dir: "azure-programmableconnectivity"
```
The package name may not be correct. It would be decided in SDK review meeting.

weidongxu-microsoft commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/common.tsp:
```diff
@@ -0,0 +1,110 @@
+
+alias Operations = Azure.Core.ResourceOperations<
+  ServiceTraits,
+  ApcErrorResponse
```
The shape of the error model is not conformable to guideline https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#rest-error-response-body-structure

Is it part of "CAMARA-like"?

weidongxu-microsoft commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/common.tsp:
```diff
@@ -0,0 +1,110 @@
+}
+
+@doc("The phone number of the device.")
+model PhoneNumberModel {
```
nit, SDK may not want to call everything "---Model"

weidongxu-microsoft commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/location.tsp:
```diff
@@ -0,0 +1,81 @@
+
+/// Endpoints
+
+@resource("DeviceLocation")
```
Path should be kebab-casing or camel-casing.

Again it may due to CAMAL.

weidongxu-microsoft commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/location.tsp:
```diff
@@ -0,0 +1,81 @@
+  @key
+  @doc("Static endpoint")
+  @visibility("read")
+  location: LocationEnum;
+}
+
+/* Needed to add this enum because typespec was creating an x-ms-enum without a name
+   which was causing a failure during linting. */
+@doc("Location static endpoint enum")
+enum LocationEnum {
+  @doc("Static endpoint")
+  location: "Location",
+}
```
Is it an Enum that later could have more values? Or is it a constant that it would NEVER have more values?

ellabrmicrosoft committed:
change endpoint casing and fix formatting

ellabrmicrosoft committed:
update examples

aronhegedusmicrosoftp committed:
add c sharp

aronhegedusmicrosoftp committed:
rename some variables because .net generatiopn is complaining about it. Request>content and response>Result

aronhegedusmicrosoftp committed:
some more replaces

aronhegedusmicrosoftp committed:
Fix single word classes:

Single word class names are too generic and have high chance of collision with BCL types or types from other libraries

aronhegedusmicrosoftp committed:
try adding veriosning

aronhegedusmicrosoftp committed:
try adding ver

ellabrmicrosoft committed:
fix typo

aronhegedusmicrosoftp committed:
fix date

aronhegedusmicrosoftp committed:
replace NetworkDevice >>> NetworkRetrieveResult

AndrewNivenMSFT committed:
Typespec validation pass

aronhegedusmicrosoftp committed:
replace older version with current one

AndrewNivenMSFT committed:
Further pipeline fixing

aronhegedusmicrosoftp committed:
use proper nouns, and do easy changes

aronhegedusmicrosoftp committed:
camel case

aronhegedusmicrosoftp committed:
replace *Interface with base path of the API

aronhegedusmicrosoftp committed:
Retrieve>Retrieval

aronhegedusmicrosoftp committed:
blankline

aronhegedusmicrosoftp committed:
rename some variables

aronhegedusmicrosoftp committed:
new lines prettier

aronhegedusmicrosoftp committed:
number interface > number verification

aronhegedusmicrosoftp committed:
new lines

aronhegedusmicrosoftp committed:
blank lines

AndrewNivenMSFT committed:
Add deprecation notice

aronhegedusmicrosoftp committed:
remove enums and add response header

aronhegedusmicrosoftp committed:
remove devicenetworkidentifier, as it's duplicate, change long/latitude to doubles

AndrewNivenMSFT committed:
Make ApcError an extension of Azure.Core.Foundations.Error

jamie-ford-microsoft committed:
Number verification updates for frontend auth

AndrewNivenMSFT committed:
ImplDocument number verification in typespec

AndrewNivenMSFT committed:
Remove python and .net sdk configuration, pending resolution of namespace discussions

AndrewNivenMSFT committed:
Update date of api version

AndrewNivenMSFT committed:
Some pipeline wrangling

AndrewNivenMSFT committed:
Remove NumberVerification retrieval APIs as they will not be implemented for public preview; add suppression of ValidResponseCodeRequired

AndrewNivenMSFT committed:
Add back Python SDK after fixes to autogen

DominikMe commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/common.tsp:
```diff
@@ -0,0 +1,117 @@
+  - IPv6 of a device in IETF 5952 format.
+  - NetworkCode matching our documentation or an output from /Network:retrieve."
+  */
+  identifier: string;
```
This seems to be missing a @doc comment

DominikMe replied:
Related https://github.com/microsoft/typespec/issues/1777

DominikMe commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/common.tsp:
```diff
@@ -0,0 +1,117 @@
+          - address/mask - an IP v6 number with a mask:
+            - 2001:db8:85a3:8d3::0/64
+            - 2001:db8:85a3:8d3::/64 */
+  ipv6: string;
```
The SDK emitter probably won't pick up the description because it's not a @doc

DominikMe replied:
Related https://github.com/microsoft/typespec/issues/1777

DominikMe commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/location.tsp:
```diff
@@ -0,0 +1,73 @@
+/// Common models
+@doc("Device information needed by operator to provide location information. Include exactly one of these properties to identify your device.")
+model LocationDevice {
+  ...NetworkAccessIdentifierModel;
```
Everything inside the spread models is optional - if they are mutually exclusive this could this be modeled as a union

AndrewNivenMSFT replied:
Unfortunately this gives the warning "Unions cannot be emitted to OpenAPI v2 unless all options are literals of the same type". I tried both standard unnamed union syntax (i.e. `alias LocationDevice = NetworkAccessIdentifierModel | PhoneNumberModel | ...`) and also using a named union decorated with @discriminator, and both of them gave me that same error. 

In an [earlier comment](https://microsoft.sharepoint.com/:w:/r/teams/CommunicationsSDK/Shared%20Documents/Alpaca%20-%20(Gateway)%20-%20General/API%20and%20SDK%20Review%20Processes/APC%20API%20questions%20.docx?d=we5538ba2edea4887a78ab582719792e9&csf=1&web=1&e=FlcGIP&nav=eyJjIjoxOTIxMzQ2Mzk1fQ%3D%3D) you suggested that it's simpler to just document the discriminated union - I've added a bit more clarity to the docstring on the locationDevice field

DominikMe replied:
Hm, the SPO link doesn't work for me. But I guess I had linked https://typespec.io/docs/standard-library/discriminated-types#using-polymorphism? The discriminator polymorphic style should work technically. We can merge it like it is now and revisit later for reevaluation in a new version.

AndrewNivenMSFT replied:
Yes, that's one of the syntaxes I tried that gave me the error above. We could suppress it, but yes, I agree better to just merge now and revisit later

DominikMe reviewed (approved):
Lgtm with minor comments

DominikMe commented on specification/programmableconnectivity/data-plane/readme.md:
```diff
@@ -0,0 +1,56 @@
+input-file:
+  - Azure.ProgrammableConnectivity/preview/2024-02-09-preview/openapi.json
+suppressions:
+  - code: ValidResponseCodeRequired
```
Had you tried adding this suppression inline in the TypeSpec to keep it local? https://azure.github.io/typespec-azure/docs/troubleshoot/suppresswarnings

AndrewNivenMSFT replied:
This is an openapi validation rather than a typespec validation, but I followed the instructions here to localise it: https://github.com/Azure/autorest/blob/main/docs/generate/suppress-warnings.md

I also didn't actually know we could suppress Typespec validation! https://eng.ms/docs/products/azure-developer-experience/design/specs-pr-guides/pr-suppressions suggests that this isn't possible so I didn't try. I've used that to suppress another error relating to returning an empty body in our 302 response.

AndrewNivenMSFT committed:
Improve docstrings, try to localize suppressions

DominikMe commented on specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/number.tsp:
```diff
@@ -0,0 +1,102 @@
+/// Response models
+
+@doc("Empty model to return as part of 302 redirect")
+model EmptyResult {
```
This can be deleted now, right?

AndrewNivenMSFT replied:
Ah yes, just missed that, thanks for catching
